feat(continuous-deploy-fingerprint): Add optional environment input for EAS updates

tharakadesilva opened the pull request:
### Linked issue

https://github.com/expo/eas-cli/issues/2195

### Additional context

https://docs.expo.dev/eas/environment-variables/#eas-update

tharakadesilva commented:
Hi @wschurman,

I'd appreciate your review on this PR. A couple of quick notes:

1. The failing CI job appears unrelated to my changes.
2. I'm still learning the approval request process - could you point me to any documentation or guidelines I should follow?

Thanks for your help!

wschurman commented:
> The failing CI job appears unrelated to my changes.

Yep, this happens for external contributions. No worries.

> I'm still learning the approval request process - could you point me to any documentation or guidelines I should follow?

I added myself as a reviewer. I should have time to review in the next few days. I think we need to ensure that both this and the [call here](https://github.com/expo/expo-github-action/blob/main/src/actions/continuous-deploy-fingerprint.ts#L176) both use the same env variables since they could be in the app.json. And then we also probably need a way to guarantee that they are the same ones as the eas build profile uses?

I'll put some more thought into this in the next couple days.

tharakadesilva commented:
> I think we need to ensure that both this and the [call here](https://github.com/expo/expo-github-action/blob/main/src/actions/continuous-deploy-fingerprint.ts#L176) both use the same env variables since they could be in the app.json.

Ah, this is something that I missed, thanks for bringing this up!

I actually haven't referenced any variables in `app.json` in my setup, which is probably why it worked in my use case that I described [here](https://github.com/expo/eas-cli/issues/2195#issuecomment-2521718431).

This brings up an interesting question though: Should the fingerprint change when a JS bundle env var changes? For variables specified in app.json (like `SENTRY_TOKEN`), aren't these only used during build time rather than update publish time?

> And then we also probably need a way to guarantee that they are the same ones as the eas build profile uses?

Based on [the documentation](https://docs.expo.dev/eas/environment-variables/#eas-build-and-workflows-job-runs), this seems to be already handled. The environment variables are made available based on the active profile, so no additional work should be needed on our end.

The main issue was with updates. With updates:

> When the --environment flag is not provided, eas update will use the .env files present in your project directory for the update job and won't use the environment variables set on the EAS servers.

This explains why in my Supabase implementation, while the environment variables were available during build time, they weren't accessible via process.env during update publishing since they weren't present during that phase.

> I'll put some more thought into this in the next couple days.

I really appreciate your feedback and time on this! And please excuse any misunderstandings on my part – I'm still quite new to this ecosystem.

wschurman commented:
> This brings up an interesting question though: Should the fingerprint change when a JS bundle env var changes? For variables specified in app.json (like SENTRY_TOKEN), aren't these only used during build time rather than update publish time?

The fingerprint runtime version is used to declare precise compatibility between builds and updates, so if something changes the build runtime version the only way to guarantee compatibility is if the update runtime version is also changed. Anything that changes the build changes the fingerprint by default, though all users have the ability to exclude things from fingerprint if they wish via the provided exclusion mechanisms to set up custom rules for excluding certain changes from changing the fingerprint.

> > And then we also probably need a way to guarantee that they are the same ones as the eas build profile uses?
> Based on the documentation, this seems to be already handled. The environment variables are made available based on the active profile, so no additional work should be needed on our end.

Sorry, I should have been more precise. I more meant guarantee/require that the "environment" string provided here matches the "environment" string in the eas.json profile. This way, if the build profile specifies an environment we can throw a warning if the "environment" flag you add isn't provided.

wschurman commented:
Asking internally how we recommend fixing this for EAS environment variables, will get back to you in a few days.

wschurman commented:
@tharakadesilva - I added some code that should theoretically load environment variables for execution of the `npx expo config` call and the `npx expo-updates fingerprint:generate` call. I haven't had time to test it yet though.

tharakadesilva commented:
Thanks @wschurman!! Let me run a few tests from my end.

tharakadesilva commented:
Hey Will, these are the changes that I did so far:

1. https://github.com/expo/expo-github-action/pull/316/commits/052de8ac524d21879788aba2cfc20f7cbab3f3b9 - `eas env:exec preview npx expo config --json --type public` broke, so I had to wrap the sub command in quotes like this: `eas env:exec preview "npx expo config --json --type public"`. 
2. https://github.com/expo/expo-github-action/pull/316/commits/f50f1e7d5a91e9b9af9b245df81664511f0cac01 - Control silent mode based on debug mode.

Now, I am stuck on:

```
[stderr] bash: line 1: npx expo config --json --type public: command not found
❌ "npx expo config --json --type public" failed
bash -c "npx expo config --json --type public" exited with non-zero code: 127
    Error: env:exec command failed.
```

From an initial glance, it looks like the `npx` command is not found when we spawn a process with `bash -c`. I tried some things using `which` and `realpath` instead of `npx`, but didn't have any luck with them either.

I also tried a `ls -la` on the action and through `eas env:exec`. These are the results:
* Without `env:exec`: `lrwxrwxrwx 1 root root 38 Dec 11 14:05 /usr/local/bin/npx -> ../lib/node_modules/npm/bin/npx-cli.js\n`
* With `env:exec`: `ls: command not found`

For sanity, since the npx symlink is under a user local bin folder, I did a `whoami` and this is the result:
* Without `env:exec`: `runner`
* With `env:exec`: `Environment variables with visibility "Plain text" and "Sensitive" loaded from the "preview" environment on EAS: BUILD_VAR, EXPO_PUBLIC_UPDATE_VAR.\n' +
    '\n' +
    'Running command: "whoami"\n' +
    '[stdout] runner\n`

So, the user hadn't changed, but the tools were no longer being found for some reason. This might be a sandbox feature of Github actions (or am I in a completely different env altogether)...

I will keep playing around with this later today. Sharing my progress in case you might know of a quick solution for this!

tharakadesilva commented:
After reading [this article](https://www.baeldung.com/linux/eval-vs-bash-c) I tried eval here instead of `bash -c` [here](https://sourcegraph.com/github.com/expo/eas-cli/-/blob/packages/eas-cli/src/commands/env/exec.ts?L114), but ended up with:

```
Environment variables with visibility "Plain text" and "Sensitive" loaded from the "preview" environment on EAS: BUILD_VAR, EXPO_PUBLIC_UPDATE_VAR.
Running command: "npx expo config --json --type public"
❌ "npx expo config --json --type public" failed
spawn eval ENOENT
    Error: env:exec command failed.
```

---

Another potential approach to this seems to be [eas env:pull](https://github.com/expo/eas-cli?tab=readme-ov-file#eas-envpull-environment). However, that doesn't solve [this exact problem](https://docs.expo.dev/eas/using-environment-variables/#use-environment-variables-for-other-commands).

--- 

Didn't find any useful [usages on Sourcegraph](https://sourcegraph.com/search?q=context:global+env:exec&patternType=keyword&sm=0) as well.

tharakadesilva commented:
Adding `shell: true` [here](https://sourcegraph.com/github.com/expo/eas-cli/-/blob/packages/eas-cli/src/commands/env/exec.ts?L114) seems to solve the issue. 🎉  I'll send a MR for it.

Ref: https://stackoverflow.com/a/44987029

There be a few more 🐉 to slay, at least I am unblocked for now.

wschurman commented on src/project.ts:
```diff
@@ -1,18 +1,36 @@
+  let args: string[];
+  if (easEnvironment) {
+    commandLine = await which('eas', true);
+    const commandToExecute = ['npx', ...baseArguments].join(' ').replace(/"/g, '\\"');
```
nice catch on the quotes, and especially nice catch on the replace!

wschurman commented:
Going to do another round of debugging/testing with a local project of mine. It seems like adding `shell: true` makes running the command in a normal prompt not work (even if it works here):

```
$ neas env:exec preview "npx expo config --json --type public"
No environment variables with visibility "Plain text" and "Sensitive" found for the "preview" environment on EAS servers.

Running command: npx expo config --json --type public
[stdout] Entering npm script environment at location:
[stdout] /Users/wschurman/temp/pkfsapjfasafjpfajaf
[stdout] Type 'exit' or ^D when finished
```

tharakadesilva commented:
Added these two commits, I am not too happy with them. 😓 

* https://github.com/expo/expo-github-action/pull/316/commits/06f12244f0110238e2105aa1994b25b8155a4050 (the stdout processing needs some "massaging" as there is an extra stdout calls to print saying that the expo env is being used and the env vars are being loaded.
* https://github.com/expo/expo-github-action/pull/316/commits/88a9191ec0e6f4e5ab5fb8c562a763415afe65d5 (not entirely sure if this is necessary, but the stdout seems to be polluted when we do Promise.all; I did this very late at night so I might have been hallucinating a bit too; I'll double check if we need this).

tharakadesilva commented:
> Going to do another round of debugging/testing with a local project of mine.

I was running a few tests as well, but I ran into an issue and I didn't get a chance to take a look at it again.

My android build works fine, but on iOS I am getting an error saying that the fingerprint has changed (similar to [this](https://github.com/expo/eas-cli/issues/2448)).

These are my builds:

* https://expo.dev/accounts/tharakadesilva/projects/template/builds/59d24f21-f7b8-4643-9d52-4f02853b2363
* https://expo.dev/accounts/tharakadesilva/projects/template/builds/122f90ab-06df-4819-9f77-594e648b0147

This is a new template project that I am working on to make my life easier, not the projects where I have `continuous-deploy-fingerprint` already working. I will try with one of those projects as well. It might be a misconfiguration in my new template project.

Sorry for the delay, I haven't gotten a chance to work on this during the week. I'll try to take a look again towards the end of the week.

tharakadesilva commented:
> It seems like adding `shell: true` makes running the command in a normal prompt not work (even if it works here):
> 
> ```
> $ neas env:exec preview "npx expo config --json --type public"
> No environment variables with visibility "Plain text" and "Sensitive" found for the "preview" environment on EAS servers.
> 
> Running command: npx expo config --json --type public
> [stdout] Entering npm script environment at location:
> [stdout] /Users/wschurman/temp/pkfsapjfasafjpfajaf
> [stdout] Type 'exit' or ^D when finished
> ```

Oof, that is not good. Maybe `spawnAsync` requires better args? Unfortunately, I am not a node expert, but if you don't find a good solution, I'll take a look at it along with the fingerprinting issue that I was seeing.

tharakadesilva commented:
Also, I am sharing my test plan for your review and feedback:

# Test Env

I started with the following vars:

<img width="1215" alt="image" src="https://github.com/user-attachments/assets/846913f8-ce82-43aa-8686-dd05e0a3486f" />

To make sure that the vars were available during build time, I created a small expo plugin:

```
const process = require('process');

const withMySDK = (config) => {
  console.log('buildVar', process.env.BUILD_VAR);
  console.log('updateVar', process.env.EXPO_PUBLIC_UPDATE_VAR);
  return config;
};

module.exports.withMySDK = withMySDK;
```

I also updated my RN components to obtain the env vars and try to show them (build var should not be available):

```
const buildVar = process.env.BUILD_VAR;
const updateVar = process.env.EXPO_PUBLIC_UPDATE_VAR;

...

<Paragraph ta='center'>buildVar: {buildVar}</Paragraph>
<Paragraph ta='center'>updateVar: {updateVar}</Paragraph>
```

# Test Summary

These are the scenarios that I want to test:

- [ ] Trigger a new pipeline (no existing builds / updates)
    - [ ] Both env vars are accessible in the build plugin
    - [ ] Can access the `EXPO_PUBLIC` env var in JS content (update part), but not the build env var
    - [ ] Fingerprints match between build and update
- [ ] Trigger a new pipeline (no changes)
    - [ ] No new builds are triggered
    - [ ] A new update is posted (fingerprint is same as the previous update)
    - [ ] Can access the `EXPO_PUBLIC` env var in JS content (update part), but not the build env var
- [ ] Modify both env vars and trigger a new pipeline
    - [ ] A new build is triggered?
    - [ ] A new update is posted
    - [ ] Both **updated** env vars are available in the build plugin
    - [ ] Can access the **updated** `EXPO_PUBLIC` env var in JS content (update part), but not the build env var
    - [ ] Fingerprints match between build and update

tharakadesilva commented:
> It seems like adding `shell: true` makes running the command in a normal prompt not work (even if it works here):
> 
> ```
> $ neas env:exec preview "npx expo config --json --type public"
> No environment variables with visibility "Plain text" and "Sensitive" found for the "preview" environment on EAS servers.
> 
> Running command: npx expo config --json --type public
> [stdout] Entering npm script environment at location:
> [stdout] /Users/wschurman/temp/pkfsapjfasafjpfajaf
> [stdout] Type 'exit' or ^D when finished
> ```

I have a degenerate idea, but I haven't tested it yet. Something like this [here](https://sourcegraph.com/github.com/expo/eas-cli/-/blob/packages/eas-cli/src/commands/env/exec.ts?L114)

```diff
diff --git a/packages/eas-cli/src/commands/env/exec.ts b/packages/eas-cli/src/commands/env/exec.ts
index 227e340c..0d376579 100644
--- a/packages/eas-cli/src/commands/env/exec.ts
+++ b/packages/eas-cli/src/commands/env/exec.ts
@@ -111,7 +111,9 @@ export default class EnvExec extends EasCommand {
     environmentVariables: Record<string, string>;
   }): Promise<void> {
     Log.log(`Running command: ${chalk.bold(command)}`);
-    const spawnPromise = spawnAsync('bash', ['-c', command], {
+    const commandParts = command.split(' ');
+    const spawnPromise = spawnAsync(commandParts[0], commandParts.slice(1), {
+      shell: true,
       stdio: ['inherit', 'pipe', 'pipe'],
       env: {
         ...process.env,
```

Not sure if this will work, but throwing out the idea before I clock out.

tharakadesilva commented:
I started with these:

![image](https://github.com/user-attachments/assets/1da4dbcc-ee6e-457c-af95-3303db4e6be8)

These are the test I ran with their observations:

1. Initial publish ([iOS build](https://expo.dev/accounts/tharakadesilva/projects/template/builds/f4778f71-b7aa-4a3e-b85b-efa7bf2a1754) | [Android build](https://expo.dev/accounts/tharakadesilva/projects/template/builds/6efa1288-fcb9-45a2-adbb-e2aa82812232) | [iOS update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/f69c88c1-d682-41ee-8bb7-6c45a2ef815e) | [Android update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/9ccf159e-0ff9-453f-8945-871216f0466b))

* New builds are created for iOS and Android.
* Updates are created.
* Update fingerprints match the build fingerprints.
* `BUILD_VAR` and `EXPO_PUBLIC_UPDATE_VAR` are available when build fingerprint is made.
* `EXPO_PUBLIC_UPDATE_VAR` is available to be used during app update, but `BUILD_VAR` is not.

2. Rerun pipeline ([iOS update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/fe39cc10-ee64-417d-a3a9-2b5ff953fe78) | [Android update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/a790ac8e-71f2-4e0f-8eb5-8996c54ef263))

* No new builds are created for iOS and Android.
* Updates are created.
* Update fingerprints match the build fingerprints.
* `BUILD_VAR` and `EXPO_PUBLIC_UPDATE_VAR` are available when build fingerprint is made.
* `EXPO_PUBLIC_UPDATE_VAR` is available to be used during app update, but `BUILD_VAR` is not.

3. Update `BUILD_VAR` to `new_build_value` ([iOS update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/0125cbe7-c8f0-48cc-af2b-2b7e96defa09) | [Android update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/5ef08802-c3a7-4b32-ae34-4799f6ed102d))

* No new builds are created for iOS and Android.
* Updates are created.
* Update fingerprints match the build fingerprints.
* `BUILD_VAR` and `EXPO_PUBLIC_UPDATE_VAR` are available when build fingerprint is made.
* `EXPO_PUBLIC_UPDATE_VAR` is available to be used during app update, but `BUILD_VAR` is not.

4. Update `EXPO_PUBLIC_UPDATE_VAR ` to `new_update_value ` ([iOS update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/b83e17ee-d7c7-40ab-be1f-f7041dd1d3d4) | [Android update](https://expo.dev/accounts/tharakadesilva/projects/template/updates/ccb91045-81e2-486d-89cc-96443fde7411))

* No new builds are created for iOS and Android.
* Updates are created.
* Update fingerprints match the build fingerprints.
* `BUILD_VAR` and `EXPO_PUBLIC_UPDATE_VAR` are available when build fingerprint is made.
* `EXPO_PUBLIC_UPDATE_VAR` is available to be used during app update, but `BUILD_VAR` is not.

wschurman commented:
Just coming back from vacation, looking at this again now.

wschurman commented:
I'm going to make some changes to the `eas env:exec` command to make it just a pass-through and not do any logging itself.

wschurman commented:
https://app.graphite.dev/github/pr/expo/eas-cli/2800/eas-cli-Silence-all-non-command-output-in-non-interactive-mode-of-eas-env-exec

tharakadesilva commented:
Thanks Will!

wschurman reviewed (changes requested):
Back to your queue to use the `--non-interactive` flag on `env:exec` as well as your PR in eas-cli once it lands. I think this should be good to go then and I'll do a full round of testing in my test repos that use this action.

tharakadesilva commented:
> Back to your queue to use the --non-interactive flag on env:exec as well as your PR in eas-cli once it lands. I think this should be good to go then and I'll do a full round of testing in my test repos that use this action.

Will address this over the weekend.

tharakadesilva commented:
> Added these two commits, I am not too happy with them. 😓

Removed these commits and switched to `--non-interactive`. Things are working like a charm! Thanks for all the help, Will.

Once this is merged I need to start rewriting git history to remove the eas-cli large files I committed for testing... 😅

wschurman reviewed (approved):
Merged in https://github.com/expo/expo-github-action/commit/8134c7a005e20406f4bdf78ecc1679c373b32a3f. Thanks for putting in the effort to get this fully working!

tharakadesilva commented:
Thanks and thanks for all the help and guidance!
